core/zoom_controls.js

```diff
@@ -194,44 +197,82 @@ Blockly.ZoomControls.prototype.dispose = function() {
   }
 };
 
+/**
+ * Returns the bounding rectangle of the UI element in pixel units relative to
+ * the Blockly injection div.
+ * @returns {Blockly.utils.Rect} The plugin’s bounding box.
+ */
+Blockly.ZoomControls.prototype.getBoundingRectangle = function() {
+  var bottom = this.top_ + this.HEIGHT_;
+  var right = this.left_ + this.WIDTH_;
+  return new Blockly.utils.Rect(this.top_, bottom, this.left_, right);
+};
+
+
 /**
  * Position the zoom controls.
  * It is positioned in the opposite corner to the corner the
  * categories/toolbox starts at.
+ * @param {!Blockly.MetricsManager.ContainerRegion} viewMetrics The workspace
+ *     viewMetrics.
+ * @param {!Blockly.MetricsManager.AbsoluteMetrics} absoluteMetrics The absolute
+ *     metrics for the workspace.
+ * @param {!Blockly.MetricsManager.ToolboxMetrics} toolboxMetrics The toolbox
+ *     metrics for the workspace.
+ * @param {!Array<Blockly.utils.Rect>} savedPositions List of rectangles that
+ *     are already on the workspace.
  */
-Blockly.ZoomControls.prototype.position = function() {
+Blockly.ZoomControls.prototype.position = function(
+    viewMetrics, absoluteMetrics, toolboxMetrics, savedPositions) {
   // Not yet initialized.
   if (!this.verticalSpacing_) {
     return;
   }
-  var metrics = this.workspace_.getMetrics();
-  if (!metrics) {
-    // There are no metrics available (workspace is probably not visible).
-    return;
-  }
-  if (metrics.toolboxPosition == Blockly.TOOLBOX_AT_LEFT ||
+  if (toolboxMetrics.position == Blockly.TOOLBOX_AT_LEFT ||
       (this.workspace_.horizontalLayout && !this.workspace_.RTL)) {
     // Toolbox starts in the left corner.
-    this.left_ = metrics.viewWidth + metrics.absoluteLeft -
+    this.left_ = viewMetrics.width + absoluteMetrics.left -
         this.WIDTH_ - this.MARGIN_SIDE_ - Blockly.Scrollbar.scrollbarThickness;
   } else {
     // Toolbox starts in the right corner.
     this.left_ = this.MARGIN_SIDE_ + Blockly.Scrollbar.scrollbarThickness;
   }
 
-  if (metrics.toolboxPosition == Blockly.TOOLBOX_AT_BOTTOM) {
-    this.top_ = this.verticalSpacing_;
+  // Upper corner placement
+  var minTop = this.top_ = this.verticalSpacing_;
+  // Bottom corner placement
+  var maxTop = viewMetrics.height + absoluteMetrics.top -
+      this.HEIGHT_ - this.verticalSpacing_;
+  var placeBottom = toolboxMetrics.position !== Blockly.TOOLBOX_AT_BOTTOM;
+  this.top_ = placeBottom ? maxTop : minTop;
+  if (placeBottom) {
+    this.zoomInGroup_.setAttribute('transform', 'translate(0, 43)');
+    this.zoomOutGroup_.setAttribute('transform', 'translate(0, 77)');
```

Comment:
Probably not for this PR because it looks like these were already there. But this looks pretty suspicious. We should probably at least create local variables and add comments so we know where these numbers are coming from.

Reply:
I left this logic unchanged. I don't fully understand what it's doing, but I think it has to do with rearranging the order from top to bottom of zoom in, zoom out, and zoom reset button.